File: log2a.py
def binLeng(bin : int):
    size = bin.__sizeof__() - 24
    maxE = (size >> 1) * 15 # same as size // 4 * 30
    
    for val in reversed(range(maxE)): 
        if bin & (1 << val): return(val+1)

def log2a(n : int):
    l = binLeng(n) #n.bit_length()
    if n == 0: return(0)
    r = (n & (1 << (l-1))-1 ) / (1<<l) *2 
    return( float(l - 1) + r)

File: log2a.c
#include <stdio.h>

const int mask = (1 << 23) -1;
const int expnt = -1 - mask;

int binLeng( unsigned int bin ) {
    for (int i = sizeof(int) * 8-1; i > 0; i--) {
        if ( ( bin & (1 << i) ) > 0) {
            return(i);
        }
    };
    return(0);
}


float log2a( int n ) {
    if (n == 0) {return(0.0F);}
    
    int l = binLeng(n);
    float r = (n & ( 1 << l  ) - 1 );
    int tmp = (( (* (int *) &r) & expnt) -( l  << 23))  | (( (* (int *) &r) & mask));
   

    return( ((float) l - 1) + * (float *) &tmp );
}
